Feature/profile settings

Davowowo opened the pull request:
feat: Fixing profile management features to settings screen

- Uploading / Deleting ProfilePictures with LoadingCircle
- Dynamicly Adapt the Picture
- Changing Name with Loading Circle
- Dynamicly Adapt the Name
- Sending a verification-link to an given email -> when verified the email is updated and the user must log in again 
- Changing Password with Authentification and Double Input 
- Logout have an Dialog Window now
- The Delete Dialog has now colors to see if the Verification is correct
- Snackbar Feedback is added

Davowowo committed:
feat: Add profile management features to settings screen

- Implemented name change functionality with dialog support.
- Added email change feature with validation.
- Introduced password change flow with re-authentication and new password validation.
- Added account deletion dialog with confirmation requirement for safety.
- Added confirmation dialog before logout

Davowowo committed:
feat: Fixing profile management features to settings screen

- Uploading / Deleting ProfilePictures with LoadingCircle
- Dynamicly Adapt the Picture
- Changing Name with Loading Circle
- Dynamicly Adapt the Name
- Sending a verification-link to an given email -> when verified the email is updated and the user must log in again
- Changing Password with Authentification and Double Input
- Logout have an Dialog Window now
- The Delete Dialog has now colors to see if the Verification is correct
- Snackbar Feedback is added

FabianRolfMatthiasNoll commented on event_radar/lib/main.dart:
```diff
@@ -118,10 +115,9 @@ class MyApp extends StatelessWidget {
-                    email: user?.email,
-                    name: user?.displayName,
+                  return ChangeNotifierProvider(
+                    create: (_) => EventCreationViewModel(),
```
Warum wird hier ein Event Creation ViewModel erstellt wenn wir zu den ProfileSettings gehen?

Davowowo replied:
das ist mir glaub untergegangen als ich copy pasted habe

FabianRolfMatthiasNoll commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,643 @@
+
+  //ChangeProfilePictureDialogWindow
+  Future<void> updateProfilePicture(BuildContext context) async {
+    final viewModel = Provider.of<EventCreationViewModel>(
```
Hier bitte keine ViewModels anderer Screens verwenden das kann für krasse Probleme sorgen. Wenn du die picker componente da drin brauchst. Dann erstelle die unter widget und zieh die da hin und an allen orten wo die verwendet wird bau die bitte ein. In diesem Fall aus EventCreation rausholen und dort im Screen dann das widget verwenden ebenso wie hier

Davowowo replied:
ich verstehe nicht ganz wie du meinst

FabianRolfMatthiasNoll replied:
Du benutzt hier das EventCreationViewModel das was du dir da in der main mitgeben hast um den imagePicker zu verwenden. Offensichtlich brauchst du die gleiche Komponente. Allerdings nimmt man dann nicht einfach das ganze ViewModel sondern man sagt sich oh ok diese Komponente wird mehrfach gebraucht also mach ich die als eigenständiges Widget wie z.B wie das password_form_field oder was map_snippet widget. Heißt du nimmst die Logik aus dem EventCreationViewModel und packst sie in ne eigene Datein unter widgets und dann kannst du überall wo du die original gebraucht hast direkt einbinden.

FabianRolfMatthiasNoll reviewed (changes requested):
Ich habs mir noch nicht live angeschaut sondern erstmal nur den Code überflogen und das ist eine wichtige änderung

Davowowo committed:
fix: fixed unnecessary interaction with eventcreation and added an image_picker.dart in utils which can be used to pick and crop an image from gallery if called.

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                        );
+                      }
+                    }
+                  },
```
Die ganze onClick Funktion würde ich auch zumindest in eine Methode dieser Klasse extrahieren, und wahrscheinlich steckt da auch ne Funktion drin, die in auth_service passt

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                      );
+                    }
+                  }
+                },
```
same here

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                          SnackBar(content: Text('Fehler: ${e.toString()}')),
+                        );
+                      }
+                    },
```
and here

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                    SnackBar(content: Text('Fehler: ${e.toString()}')),
+                  );
+                }
+              },
```
and here

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
-                  ? Text(getImagePlaceholder(title))
-                  : null,
+        GestureDetector(
+          onTap: () => updateProfilePicture(context),
```
Wenns kein speziellen Grund dagegen gibt würd ich immer den Inkwell benutzen um etwas klickbar zu machen. Gibt dem Nutzer besseres Feedback

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                                ),
+                              );
+                            }
+                          }
```
same here

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                    SnackBar(content: Text('Fehler: ${e.toString()}')),
+                  );
+                }
+              },
```
Das hier ist eigentlich Geschäftslogik und sollte nicht in der UI sein. Alles wo FirebaseAuth verwendet würde ich nach auth_service verschieben

FabianRolfMatthiasNoll commented:
Jetzt hast du ja die Logik extrahiert. Jetzt macht es ja auch Sinn die Logik aus dem EventCreationViewModel zu entfernen und dein neues Widget dort einzusetzen

MaarxS reviewed (changes requested):
Hab jetzt nicht alles komplett angeschaut, aber eine Anmerkung, die ich hab: Ich würde die UI und Geschäftslogik streng trennen. Ich wäre für folgende Regeln:
- Wenn eine anonyme Funktion (also zum Beispiel bei onPressed) mehr als ein Statement hat, dann würd ich die in eine Methode der Klasse verschieben.
- Wenn in einer Funktion etwas externes verwendet wird (also z.B. FirebaseAuth), würd ich das in eine Service Klasse verschieben.

Weiß nicht, ob ich alles gefunden hab, am besten einfach mal mit strg + F nach onPressed und onClick und onTap suchen und nochmal schauen, ob es sinnvoll ist.

Funktionalität muss ich auch noch ausprobieren.

MaarxS commented:
Ahh jetzt check ich endlich, wie diese Code preview funktioniert, der nimmt immer die letzten vier Zeilen. Ich hatte gehofft, dass der bis zur öffnenden Klammer alles nimmt. Also die Kommentare beziehen sich eigentlich alle auf die gesamte anonyme Funktion.

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                            onPressed: () {
+                              Navigator.of(context).pop();
+                              AuthService().signOut();
+                              context.go('/login/reset-password');
```
Finde das bisschen verwirrend, dass man mit dem "Passwort vergessen" Button ausgeloggt wird. Ich würde eher mit context.push zum passwort reset gehen, oder einfach direkt die passwort reset mail schicken, weil die Email um dies geht ja schon klar ist.

Falls das mit dem Router unklar war, wenn man go aufruft, dann wird der stack gecleared und stattdessen mit den Elementen vor dem slash gefüllt. Wenn man push aufruft kommt der aktuelle Pfad auf den Stack und das Ziel wird aufgerufen.

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
       children: [
         InkWell(
           onTap: () {},
           child: Container(
```
Bei dem InkWell müsste man noch die onTap Funktion ausfüllen, oder ihn entfernen. Ich hatte den glaube ich ursprünglich gedacht auch um den Name zu ändern, ist aber eigentlich nicth nötig, wenn wir den IconButton haben. Geschmackssache.

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+  const ProfileSettingsScreen({super.key});
+
+  @override
+  _ProfileSettingsScreenState createState() => _ProfileSettingsScreenState();
```
Hier `State<ProfileSettingsScreen>` zurückgeben

MaarxS commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -1,115 +1,639 @@
+                } catch (e) {
+                  ScaffoldMessenger.of(context).showSnackBar(
+                    SnackBar(content: Text('Fehler: ${e.toString()}')),
+                  );
```
Hier und weiter oben überprüfen, ob der context noch mounted ist mit `if (context.mounted)` weil sonst kanns sein, dass der Nutzer schon weg navigiert hat. Sollte man auch überall machen, wo die Warning angezeigt wird.

Sehr offensichtlich wird das, wenn man ganz oft ne ungültige Email angiebt, weil dann kommen die Snackbars immer wieder, selbst wenn man wegnavigiert.

Man sieht die Snackbars auch nicht so wirklich gut hinter dem AlertDialog, vielleicht gibts da ne bessere Lösung, wie das Textfeld rot markieren, oder ein weiterer AlertDialog, weiß nicht, was da gut kommt. Sieh das mal als optional an.

MaarxS reviewed (changes requested):
So habs mir jetzt mal angeschaut, sorry, dass ich so viel zum Meckern hab 😅 Aber die Funktionalität sieht gut aus, da hast du ja schon viel implementiert.

FabianRolfMatthiasNoll commented:
Mir ist leider auch noch etwas eingefallen was hier definitv in die Funktionalität gehört. Und zwar eine Cloud Function die aufgerufen wird wenn man einen User löscht der dann sämtliche Sachen wie Profilbild löscht, Oder einträge in der participants list eines Events

Davowowo committed:
Add .Idea to gitignore

Davowowo committed:
Merge remote-tracking branch 'origin/develop' into feature/profileSettings

# Conflicts:
#	event_radar/lib/views/profile/profile_settings_screen.dart

Davowowo committed:
Extract image_picker logic from event_creation_viewmodel.dart

Davowowo committed:
feat: adding ForgotPassword Fuctionality and fixing comments to pull request

FabianRolfMatthiasNoll commented on event_radar/lib/views/profile/profile_settings_screen.dart:
```diff
@@ -153,26 +173,20 @@ class _ProfileSettingsScreenState extends State<ProfileSettingsScreen> {
+                          ScaffoldMessenger.of(context).showSnackBar(
+                            SnackBar(
+                              content: Text(
+                                'Das eingegebene Passwort ist invalide, versuchen Sie es erneut.',
```
invalide? Es ist behindert? Ich würde sagen das es falsch ist oder? :D

Davowowo replied:
naja es kann auch sein dass es ein Timeout gab -> es muss unbedingt falsch gewesen sein
Die Klausel, fängt glaube ich 3 Cases ab, aber ja falsch ist der häufigste Fall

FabianRolfMatthiasNoll replied:
Ok aber dann ist die Benennung Invalide doch trotzdem falsch das bedeutet doch im deutschen etwas völlig anderes oder bin ich jetzt dumm?

FabianRolfMatthiasNoll commented on event_radar/lib/views/profile/reset_password_screen.dart:
```diff
@@ -1,78 +1,125 @@
       return 'Bitte Email eingeben';
     }
+    final emailRegex = RegExp(
+      r'^[a-zA-Z0-9._%-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,4}$',
```
Regex doesnt match for example google extension format like
Fabian.Noll+Github@googlemail.de

FabianRolfMatthiasNoll replied:
So Email sachen sind meist auch bereits von anderen gelöst worden macht meistens sinn das einfach zu benutzen
https://pub.dev/packages/email_validator

Davowowo committed:
feat: saving many changes to merge develop into my branch to avoid event polling

Davowowo committed:
Merge remote-tracking branch 'refs/remotes/origin/develop' into feature/profileSettings

# Conflicts:
#	event_radar/lib/core/viewmodels/event_creation_viewmodel.dart
#	event_radar/lib/main.dart
#	event_radar/lib/views/event_creation_screen.dart
#	event_radar/lib/views/profile/profile_settings_screen.dart

Davowowo committed:
feat: fixing Bugs

Davowowo committed:
feat: fixing Bugs

Davowowo committed:
feat: visual update for user experience and fixing the red line, fixing pending Email Use across multiple accounts

Davowowo committed:
feat: final push for profile_settings_screen.dart
now it is possible for the user to reload to start a new session due the email-change when he refreshes the site. Even with the link that resets your account to the previous email.

Davowowo committed:
feat: invalid Password gone, refactoring, nice ui responses with grammar
